module/modmanager/manager.go

```diff
@@ -595,23 +594,31 @@ func (mgr *Manager) ValidateConfig(ctx context.Context, conf resource.Config) ([
 	return resp.Dependencies, nil
 }
 
-// ResolveImplicitDependenciesInConfig modifies the config diff to add implicit dependencies to changed resources
-// and updates modular resources whose module has been changed with new implicit deps and adds them to conf.Modified.
+// ResolveImplicitDependenciesInConfig mutates the passed in diff to add modular implicit dependencies to added
+// and modified resources. It also puts modular resources whose module has been modified in conf.Added if they
+// are not already there.
 func (mgr *Manager) ResolveImplicitDependenciesInConfig(ctx context.Context, conf *config.Diff) error {
 	mgr.mu.RLock()
 	defer mgr.mu.RUnlock()
-	// Find all components/services that are unmodified but whose module has been updated and add them to the modified array.
+	// Find all components/services that are unmodified but whose module has been updated and add them to conf.Added.
 	for _, c := range conf.Right.Components {
-		// See if this component is being provided by a module
+		// See if this component is being provided by a module.
 		mod, ok := mgr.getModule(c)
 		if !ok {
 			continue
 		}
 		// If it is, check against the modified modules to determine if the config should also be updated but won't already be
 		if slices.ContainsFunc(conf.Modified.Modules, func(elem config.Module) bool { return elem.Name == mod.cfg.Name }) &&
-			!slices.ContainsFunc(conf.Added.Components, func(elem resource.Config) bool { return elem.Name == c.Name }) &&
-			!slices.ContainsFunc(conf.Modified.Components, func(elem resource.Config) bool { return elem.Name == c.Name }) {
-			conf.Modified.Components = append(conf.Modified.Components, c)
+			!slices.ContainsFunc(conf.Added.Components, func(elem resource.Config) bool { return elem.Name == c.Name }) {
+			// If component is in conf.Modified, the user modified a module and its component at the same time. Remove that
+			// resource from conf.Modified and put it in conf.Added so the restarted module receives an AddResourceRequest
+			// and not a ReconfigureResourceRequest.
```

Comment:
Trying to cover an edge case here and in the analogous code for services. Let me know if my comment doesn't make sense.

Reply:
Thanks for the explanation here! I definitely agree that the conditional logic is both tough to read and your final `into` pseudocode ensures a better invariant. Tried to capture what you're describing here and added some more comments, let me know what you think.